docs/repo/dependencies-node-roadmap.md

```diff
@@ -57,15 +65,47 @@ The [`GroupedByTargetTreeViewProvider`][8] traverses down the existing `IProject
+
+At the same time the user can only interact with these items in a very limited way--there's no way for them to delete an individual assembly brought in by a NuGet package for example.
+
+These requirements lead us to represent transitive dependencies as `GraphNodes`.
+
+### DependenciesGraphProvider
 
-TODO: Describe the role and implementation of the [`IDependency`][4]`.Id` property.
+The primary connection point between the Graph Nodes and the Project System is the [`DependenciesGraphProvider`][22] class, via the `IGraphProvider` interface. It is directly responsible for the following:
 
-## Graph Nodes
+1. Specifying which graph operations it supports, via the `GetCommands` property. The only supported standard command is `GraphCommandDefinition.Contains` which is used to find the children of a given graph node.
+2. Delegate the actual implementation of graph operations to various [`IDependenciesGraphActionHandler`][23]s via the `BeginGetGraphData`/`BeginGetGraphDataAsync` methods.
+3. Handle the low-level mechanics of adding and removing nodes from the graph via its [`IDependenciesGraphBuilder`](24) implementation.
 
-Much of this document only applies to the assemblies, packages, analyzers, etc. that are _directly_ part of the project. These make up the first level of items under the Assemblies, NuGet, Analyzers, Projects, etc. nodes under Dependencies. The _transitive_ dependencies shown under that first level are not implemented as `IProjectTree` instances, but rather as graph nodes.
+### Generating new `GraphNode`s
 
-TODO: Describe the motivation behind and implementation of graph nodes for representing transitive dependencies.
+New `GraphNode`s are added to the graph as a result of operations initiated by the user. For example, when the user expands a node representing a NuGet package:
+
+1. The `IGraphProvider.BeginGetGraphData` implementation in [`DependenciesGraphProvider`][22] is called with an `IGraphContext` describing the current graph, the input node (e.g. the node for the NuGet Package) and the operation (e.g. "get the children of the input node).
+2. Each implementation of [`IDependenciesGraphActionHandler`][23] that can handle the request is asked to do so.
+3. We retrieve the current [`IDependenciesSnapshot`][6] for the project as well as the [`IDependency`][4] for the input node (via [`IAggregateDependenciesSnapshotProvider`][10]/[`IDependenciesSnapshotProvider`][2]).
+4. We find the first [`IDependenciesGraphViewProvider`][25] that supports the given [`IDependency`][4] and ask it to build up the corresponding parts of the graph.
+5. The [`IDependenciesGraphViewProvider`][25] decides what nodes to add to the graph, and calls [`IDependenciesGraphBuilder`][24]`.AddGraphNode` (implemented by [`DependenciesGraphProvider`][22]) to handle the actual mechanics.
```

Comment:
- Very important detail that is not obvious is how GraphNodes are mapped to CPS IProjectTree nodes. It is done by GraphActionHandlerBase.GetDependency method.  IGraph Provider receives all kind of nodes .cs files Properties etc and we need to correctly identify a Dependency type of nodes. At the very beginning we identify CPS nodes by it's id serialized in the file path, but then when CPS node was found first time we add to it;s GraphNode ID property to simplify identification next time we process this node. This connection puts some restriction on what can be put into FilePath of IProjectTree and IProjectItemTree nodes.
- Also please notice, that now GraphNOdes can display any custom properties in Properties panel. I added a change to GraphNodes code sometime back to support it by providing extensions vis IGraphProvider.GetExtension method. We never ended up adding them to Dependencies code because of priorities at that time, but support for that was checked in in band. If you ever need to use it ping me and i will find related code changes.

Reply:
@abpiskunov Yes, I was going to ask about your first point. When we create an `IProjectTree` instance for a NuGet package (for example) does CPS automatically create an equivalent `GraphNode` for us? I think that's what you're saying, but just want to be sure.

Reply:
CPS calls progression API (graph core logic) to notify and "map" a tree node to some graph node when tree node has ProjectTreeFlags.Common.ResolvedReference flag. Progression manages graph nodes accordingly. CPS does not explicitly kn ow anything about GraphNode objects.

Reply:
Ah, ok. My main point, though, is that these are graph nodes that represent parts of the dependencies graph, but _aren't_ created by the code in dotnet/project-system, correct?

Reply:
Yes, we are not deleting or creating those top level GraphNodes, however we are changing their properties. We are managing everything below them, though if IProjectTree node is deleted, CPS signals to progression and progression deletes the whole tree under corresponding top level GraphNode eventually, that's why we have weak collection for expanded IGraphContexts to not hold them when they are deleted.